### PR TITLE
Support taking UDT as pandas_udf input in Spark

### PR DESCRIPTION
**Pull request by da-liii**
### What changes were proposed in this pull request?
+ [x] Improve readability of serialize.py
+ [x] scalafmt/scaladoc
+ [x] update ticket ID, please update the title of https://github.com/apache/spark/pull/31735
+ [x] refactor plainSchema and add unit tests
+ [x] unit tests for plainSchema
+ [ ] POSTPONED <del>unit tests with timestamp type</del>

### How was this patch tested?
```
./build/mvn -DskipTests clean package
python/run-tests --testnames pyspark.sql.tests.test_pandas_udf_scalar
```

with the following python 3.8 env:
```
certifi==2020.12.5
coverage==5.5
numpy==1.20.1
pandas==1.2.3
pyarrow==2.0.0
python-dateutil==2.8.1
pytz==2021.1
scipy==1.6.1
six==1.15.0
xmlrunner==1.7.7
```

### CODE REVIEW

**Comment by eddyxu** [comment]
Great， this is a much better improvement 👍

File: python/pyspark/sql/pandas/serializers.py [hunk 1]
```diff
@@ -155,19 +155,14 @@ def _create_batch(self, series):
...
-        series = ((s, None) if not isinstance(s, (list, tuple)) else s for s in series)
 
-        def create_array(s, dt: DataType, t: pa.DataType):
-            mask = s.isnull()
-            if isinstance(dt, UserDefinedType):
-                s = s.apply(dt.serialize)
-            elif isinstance(dt, ArrayType) and isinstance(dt.elementType, UserDefinedType):
-                udt = dt.elementType
-                s = s.apply(lambda x: [udt.serialize(f) for f in x])
+        def create_array(s, t: pa.DataType, dt: DataType = None):
```
Comment (eddyxu) [inline python/pyspark/sql/pandas/serializers.py]: Nit: if `dt` can be none, should we mark it as `Optional[DataType]`?

File: python/pyspark/sql/pandas/serializers.py [hunk 2]
```diff
@@ -194,36 +191,61 @@ def create_array(s, dt: DataType, t: pa.DataType):
...
 
-        arrs = []
-        for s, dt in series:
-            t = to_arrow_type(dt) if isinstance(dt, DataType) else dt
-            if t is not None and pa.types.is_struct(t) and not isinstance(dt, UserDefinedType):
-                if not isinstance(s, pd.DataFrame):
-                    raise ValueError("A field of type StructType expects a pandas.DataFrame, "
-                                     "but got: %s" % str(type(s)))
+        def create_arrs_names(s, t: pa.DataType, dt: DataType = None):
+            # Input partition and result pandas.DataFrame empty, make empty Arrays with struct
```
Comment (eddyxu) [inline python/pyspark/sql/pandas/serializers.py]: Nit: `Input partition and result pandas.DataFrame empty` <- could you rephrase this setence?

**Comment by da-liii** [comment]
@eddyxu  Should be OK now. Last week's work is not well tested.

This PR is a refactor-only PR:
+ Adding tests
+ Refactor the previous PR

For maropu's review comment:

> Does it work correctly when UserDefinedType.sqlType has timestamp/dict? Could you add some tests for these types, too?

Just confirmed, the current impl does not support timestamp. Conversion of (Pandas -> Arrow -> Spark Internal Row) is too tricky. I need more time (digging into pandas and arrow) to figure out the correct way to make it work smoothly.

File: python/pyspark/sql/pandas/serializers.py [hunk 3]
```diff
@@ -155,19 +157,14 @@ def _create_batch(self, series):
...
-        def create_array(s, dt: DataType, t: pa.DataType):
-            mask = s.isnull()
-            if isinstance(dt, UserDefinedType):
-                s = s.apply(dt.serialize)
-            elif isinstance(dt, ArrayType) and isinstance(dt.elementType, UserDefinedType):
-                udt = dt.elementType
-                s = s.apply(lambda x: [udt.serialize(f) for f in x])
+        def create_array(s, t: pa.DataType, dt: Optional[DataType] = None):
+            if dt is not None:
+                if isinstance(dt, UserDefinedType):
```
Comment (da-liii) [inline python/pyspark/sql/pandas/serializers.py:163]: this part together with `pa.StructArray.from_pandas` is too tricky for me. I need to learn more backgrounds of Arrow and Pandas.

File: python/pyspark/sql/pandas/serializers.py [hunk 4]
```diff
@@ -194,36 +193,59 @@ def create_array(s, dt: DataType, t: pa.DataType):
...
+
+            if dt is None:
+                if by_field_name:
+                    return [(create_array(s[field.name], field.type), field.name) for field in t]
+                else: 
+                    return [
+                        (create_array(s[s.columns[i]], field.type), field.name)
+                        for i, field in enumerate(t)
+                    ]
+            else:
```
Comment (da-liii) [inline python/pyspark/sql/pandas/serializers.py:216]: `zip(t, dt.fields)` should only be used when t and dt is not the same.

when t = dt, just set dt to None.

**Comment by eddyxu** [comment]
> @eddyxu Should be OK now. Last week's work is not well tested.
> 
> This PR is a refactor-only PR:
> 
> * Adding tests
> * Refactor the previous PR
> 
> For maropu's review comment:
> 
> > Does it work correctly when UserDefinedType.sqlType has timestamp/dict? Could you add some tests for these types, too?
> 
> Just confirmed, the current impl does not support timestamp. Conversion of (Pandas -> Arrow -> Spark Internal Row) is too tricky. I need more time (digging into pandas and arrow) to figure out the correct way to make it work smoothly.

Yes, the current implementation does not explicitly consider `timestamp` I am not certain that they are not well supported.

**Comment by eddyxu** [comment]
Do you want to merge this PR now and left the rest to the community to review?

**Review by eddyxu — APPROVED** [review]

**Comment by da-liii** [comment]
> Do you want to merge this PR now and left the rest to the community to review?

Yes, merge it and let the community to review.

**Comment by eddyxu** [comment]
Thanks @sadhen !

**Comment by da-liii** [comment]
@eddyxu  The title should be SPARK-34799, see
 
https://issues.apache.org/jira/browse/SPARK-34799
